### PR TITLE
fix: handle case where LearnerCreditEnterpriseCourseEnrollment already exists

### PR DESCRIPTION
**Pull request by pwnage101**
When calling enroll_learners_in_courses (a.k.a. the bulk enrollment endpoint), gracefully handle the case when a subsidized enrollment (for learner credit 2) is being re-created after being revoked previously.

Before this change, the existing LearnerCreditEnterpriseCourseEnrollment object would not be found (due to transaction_id changing), then the view would fatally fail to create a new one due to a unique constraint on the `enterprise_course_enrollment_id` field.  After this change, it finds the existing object and updates a couple fields on it to re-use it.

ENT-7204

### CODE REVIEW

File: enterprise/utils.py [hunk 1]
```diff
@@ -1822,10 +1822,17 @@ def customer_admin_enroll_user_with_status(
...
+            subsidy_enrollment_obj, created = subsidized_enterprise_course_enrollment_model().objects.get_or_create(
                 enterprise_course_enrollment=obj,
+                defaults={"transaction_id": transaction_id},
             )
+            if not created:
+                # It's possible that the subsidized enterprise course enrollment already existed, whose
+                # ``transaction_id`` would correspond to a reversed ledgered transaction, and ``is_revoked`` would
+                # probably be True.  All that's in the past now, so we should override those two fields.
+                subsidy_enrollment_obj.transaction_id = transaction_id
+                subsidy_enrollment_obj.is_revoked = False
```
Comment (alex-sheehan-edx) [inline enterprise/utils.py:1834]: should we set the fulfillment's enterprise enrollment to `unenrolled=False` ?

File: enterprise/utils.py [hunk 2]
```diff
@@ -1822,10 +1822,17 @@ def customer_admin_enroll_user_with_status(
...
-                transaction_id=transaction_id,
+            subsidy_enrollment_obj, created = subsidized_enterprise_course_enrollment_model().objects.get_or_create(
                 enterprise_course_enrollment=obj,
+                defaults={"transaction_id": transaction_id},
             )
+            if not created:
+                # It's possible that the subsidized enterprise course enrollment already existed, whose
+                # ``transaction_id`` would correspond to a reversed ledgered transaction, and ``is_revoked`` would
+                # probably be True.  All that's in the past now, so we should override those two fields.
+                subsidy_enrollment_obj.transaction_id = transaction_id
```
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply enterprise/utils.py:1833]: (this is more of a thought than a request)

**Review by alex-sheehan-edx — COMMENTED** [review]
Would love a test for this, one tiny nit and an open question, but otherwise looks pretty good to me.

File: enterprise/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pwnage101) [reply enterprise/utils.py:1833]: We do have django simple history enabled on this model.  Maybe that's enough, at least for the case of manual debugging and introspection?

File: enterprise/api/v1/views.py [hunk 3]
```diff
@@ -771,7 +771,7 @@ def cancel_enrollment(self, request, fulfillment_source_uuid):
         except Exception as exc:  # pylint: disable=broad-except
             msg = (
                 f'Subsidized enrollment terminations error: unable to unenroll User {username}'
-                f'from Course {enrollment.course_id}  because: {str(exc)}'
+                f'from Course {enrollment.enterprise_course_enrollment.course_id}  because: {str(exc)}'
```
Comment (pwnage101) [inline enterprise/api/v1/views.py:774]: This change wasn't actually needed/related to this PR, but I did run into it.  The enrollment here has no course_id attribute.

File: enterprise/models.py [hunk 4]
```diff
@@ -2083,13 +2083,27 @@ def revoke(self):
...
-        self.enterprise_course_enrollment.save()
+        self.save()
 
-        # TODO revoke entitlements as well?
+    def renew(self, **kwargs):
+        """
+        Idmpotently reactivates this enterprise fulfillment source.
+        """
+        if self.enterprise_course_enrollment:
+            self.enterprise_course_enrollment.saved_for_later = False
```
Comment (pwnage101) [inline enterprise/models.py:2102]: Really not sure what saved_for_later means, but I threw it in.  Reviewers, please let me know if this is correct.

File: enterprise/settings/test.py [hunk 5]
```diff
@@ -170,6 +170,7 @@ def root(*args):
         'service_user': SERVICE_USER_THROTTLE_RATE,
     },
     'DATETIME_FORMAT': '%Y-%m-%dT%H:%M:%SZ',
+    'TEST_REQUEST_DEFAULT_FORMAT': 'json',
```
Comment (pwnage101) [inline enterprise/settings/test.py:173]: Here's another bonus change that allows all of our client.post() calls in unit tests to be cleaner.

File: test_utils/factories.py [hunk 6]
```diff
@@ -393,6 +395,7 @@ class Meta:
     transaction_id = factory.LazyAttribute(lambda x: FAKER.uuid4())
     enterprise_course_enrollment = factory.SubFactory(EnterpriseCourseEnrollmentFactory)
     is_revoked = False
+    fulfillment_type = FulfillmentTypes.LEARNER_CREDIT
```
Comment (pwnage101) [inline test_utils/factories.py:398]: Another set of unrelated changes here.  I just noticed that all the `LearnerCreditEnterpriseCourseEnrollment` objects in unit tests had the wrong fulfillment_type.

File: tests/test_enterprise/api/test_views.py [hunk 7]
```diff
@@ -4436,6 +4427,84 @@ def test_bulk_enrollment_in_bulk_courses_nonexisting_user_id(
...
+    @mock.patch("enterprise.utils.lms_enroll_user_in_course")
+    def test_bulk_enrollment_enroll_after_cancel(
+        self,
+        mock_platform_enrollment,
+        mock_get_course_mode,
+        mock_enrollment_api,
+    ):
+        """
+        Test that even after a cancelled enterprise enrollment, an attempt to re-enroll the same learner in content
+        results in expected state and payload.
```
Comment (pwnage101) [inline tests/test_enterprise/api/test_views.py:4442]: High level architecture of this test: I use factories to create the enrollment initially, then use 2 APIs to subsequently cancel then re-enroll.

File: enterprise/models.py (hunk 4 above)
Comment (iloveagent57) [reply enterprise/models.py:2102]: `saved_for_later` does basically what it says - when true, the enrollment is rendered in the "Saved for later" section of the learner portal dashboard.  I would probably just leave the value unchanged during renewal, but it's not a big deal either way.

File: enterprise/api/v1/views.py (hunk 3 above)
Comment (iloveagent57) [reply enterprise/api/v1/views.py:774]: nit: `enrollment` seems like not the right variable name in here - isn't it really a fulfillment?

File: enterprise/models.py [hunk 8]
```diff
@@ -2111,6 +2125,13 @@ class LearnerCreditEnterpriseCourseEnrollment(EnterpriseFulfillmentSource):
     .. no_pii:
     """
 
+    def renew(self, transaction_id=None, **kwargs):
+        """
+        Idmpotently reactivates this LearnerCreditEnterpriseCourseEnrollment.
+        """
+        self.transaction_id = transaction_id
```
Comment (iloveagent57) [inline enterprise/models.py:2132]: It's probably worth throwing an exception if the transaction id hasn't changed from the original fulfillment to the renewed fulfillment.  Or at least log a warning, maybe?

**Review by iloveagent57 — COMMENTED** [review]
This looks great, just one open suggestion/question of what to do if the transaction_id hasn't changed during renewal.

Also, consider using the verb `reenroll` (or "refulfill", "reactivate", whatevs) to help disambiguate from enterprise subscription renewals.

File: enterprise/api/v1/views.py (hunk 3 above)
Comment (pwnage101) [reply enterprise/api/v1/views.py:774]: totally, i can change that too

**Review by iloveagent57 — APPROVED** [review]

File: enterprise/settings/test.py (hunk 5 above)
Comment (pwnage101) [reply enterprise/settings/test.py]: backing out this change since it had the side effect of making PUT requests use JSON when they shouldn't.
